File: S5/Parallel & Distributed Computing/Lecture1:MPI/ping_pong.c
#include <mpi.h>
#include <stdio.h>

int main(int argc, char** argv) {
    MPI_Init(NULL, NULL);
    int world_size;
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);
    int world_rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);

    if (world_rank==0)
    {
        int value_to_be_sent = 41; int dest = 1;
        MPI_Send(&value_to_be_sent, 1, MPI_INTEGER, dest, 0, MPI_COMM_WORLD);

        int source = dest; int received_value=0;
        MPI_Recv(&received_value, 1, MPI_INTEGER, source, 0, MPI_COMM_WORLD, NULL);

        printf("The received answer is %d\n", received_value);
    }
    else if (world_rank==1)
    {
        int received_value = 0; int source = 0;
        MPI_Recv(&received_value, 1, MPI_INTEGER, source, 0, MPI_COMM_WORLD, NULL);

        int new_value = received_value+1;
        int dest = source;
        MPI_Send(&new_value, 1, MPI_INTEGER, dest, 0, MPI_COMM_WORLD);        
    }

    MPI_Finalize();
}

File: S5/Parallel & Distributed Computing/Lecture1:MPI/ping_pong.py
#!/usr/bin/env python
from mpi4py import MPI
import sys

world_size = MPI.COMM_WORLD.Get_size()
world_rank = MPI.COMM_WORLD.Get_rank()

comm = MPI.COMM_WORLD

if world_rank==0:
    value_to_be_sent = 41
    dest = 1
    comm.send(value_to_be_sent, dest=dest, tag=0)
    source = dest
    received_value = 0
    for _ in range(2):
        received_value = comm.recv(source=source, tag=0)
        print("The received answer is %d" % received_value)
        new_value = received_value + 1
        comm.send(new_value, dest=dest, tag=0)
    
elif world_rank==1:
    received_value = 0
    source = 0
    dest = source
    for _ in range(3):
        received_value = comm.recv(source=source, tag=0)
        if received_value != 41:
            print("The received answer is %d" % received_value)
        new_value = received_value+1
        comm.send(new_value, dest=dest, tag=0)

